tensorflow/core/common_runtime/sycl/sycl_allocator.h

```diff
@@ -46,8 +50,17 @@ class SYCLAllocator : public Allocator {
   // AllocatedSize(void* ptr) by default.
   size_t RequestedSize(void* ptr) override;
   Eigen::SyclDevice* getSyclDevice() { return sycl_device_; }
+  // Clear the SYCL device used by the Allocator
+  void ClearSYCLDevice() {
+    if(sycl_device_) {
+      mutex_lock lock(mu_);
```

Comment:
This only helps if all the reads and uses of sycl_device_ are guarded by the mutex. For example 

Thread 1 calls, e.g., Synchronize, reads non-zero sycl_device_ in  line 40.
Thread 2 calls ClearSYCLDevice, deletes the object, sets sycl_device_ = nullptr;
Thread 1 Tries to call sycl_device_->synchronize()
Boom!

I'm not sure if this is a scenario that we need to guard against?